### PR TITLE
GH2099: Cache compiled script on disk

### PR DESCRIPTION
**Pull request by tstewart65**
This pull request is what started RFC-02.  At the time I was not able to submit a pull request for the changes.

**Differences from RFC-02**

1. The cache argument was removed.  A new section was added to the cake.config file [Cache] with 2 possible properties Enabled and Path.  Enabled defaults to false and Path defaults cache under the tools folder.
2. I originally used a time stamp to determine if the cache was valid or not.  This worked for the way I was using Cake but would not work for all cases.  I have changed this to use a hash like pull request 2584 used.  I did change it to use SHA1CryptoServiceProvider instead of MD5 since MD5 is not FIPS compliant.

I have been using several versions of Cake with this  applied and have had no issues.

@mholo65 posted this in the RFC Comments
"Caching compilation is not alone enough. We must also cache registered tools resolved via #tool directive. Addins (and their dependencies) must also be loaded into app domain / load context in same order as the first time when resolved via #addin directive."

For the way this is implemented I feel just caching the compilation is enough.  I need all of the lines of the script in order to do the hash for determining if anything has changed.  The tools and addins are still being found like usual.  The main changes were made in RoslynScriptSession and in there it just determines if it needs to compile the script or run the cached DLL.

### CODE REVIEW

**Comment by mholo65** [comment]
Related to #2584 

@tstewart65 do you possible have some numbers on the performance improvements? Would be nice to see how much time we save with just caching the compilation.

**Comment by tstewart65** [comment]
Results from a very simple cake script that exists of one task and all it does it print out "Hello".
No Cache - 4.008 seconds
Caches     - 1.372 seconds

Note: Test are an average of 6 runs with tools and nugget packages already installed.

Next simple test that shows even more saving and uses a version of Cake.Recipe.  The test consists of 3 simple tasks but two of the tasks uses the RequireTool call tools.cake.  I made a change to the RequiresTool call to use a unique name for each script instead of a GUID so this script could be cached also.
No Cache - 16.493 seconds
Cached     - 6.000 Seconds

Real world example is a project that I am currently working on.
No Cache - 31.923 Seconds
Cached     - 19.287 Seconds

Below is the Verbose output from the two different basic tests no cache and cached.

Preparing to run build script...
Running Cake bootstrap...
Running build script...
Analyzing build script...
Processing build script...
Compiling build script...
**Script compile time: 1922ms**

========================================
Hello
========================================
Executing task: Hello
Hello
Finished executing task: Hello

========================================
Default
========================================
Executing task: Default
Finished executing task: Default

Task                          Duration
--------------------------------------------------
Hello                         00:00:00.0117959
Default                       00:00:00.0042251
--------------------------------------------------
Total:                        00:00:00.0160210
**build.cake execution time: 2643ms**

(Last Command - 4 seconds, 13 milliseconds)

Preparing to run build script...
Running Cake bootstrap...
Running build script...
Analyzing build script...
Processing build script...
build.cake.dll
**Running cached build script...**

========================================
Hello
========================================
Executing task: Hello
Hello
Finished executing task: Hello

========================================
Default
========================================
Executing task: Default
Finished executing task: Default

Task                          Duration
--------------------------------------------------
Hello                         00:00:00.0122724
Default                       00:00:00.0055681
--------------------------------------------------
Total:                        00:00:00.0178405
**build.cake execution time: 53ms**

(Last Command - 1 second, 389 milliseconds)

Preparing to run build script...
Running Cake bootstrap...
Running build script...
Analyzing build script...
Processing build script...
Installing addins...
build.cake.dll
Cache check failed.
Compiling build script...
**Script compile time: 2406ms**

----------------------------------------
Setup
----------------------------------------
Executing custom setup action...
  ____         _           _____             _
 / ___|  __ _ | | __  ___ |_   _|  ___  ___ | |_
| |     / _` || |/ / / _ \  | |   / _ \/ __|| __|
| |___ | (_| ||   < |  __/  | |  |  __/\__ \| |_
 \____| \__,_||_|\_\ \___|  |_|   \___||___/ \__|


Starting Setup...
Cleaning Project

========================================
Hello
========================================
Executing task: Hello
Hello
Finished executing task: Hello

========================================
VersionTest
========================================
Executing task: VersionTest
Analyzing build script...
Processing build script...
Installing tools...
Compiling build script...
**Script compile time: 2599ms**
**GitVersion.CommandLine_version=5.0.1.cake execution time: 3594ms**
Deleting file D:/CacheTest/Basic_Recipe/GitVersion.CommandLine_version=5.0.1.cake
Version
Finished executing task: VersionTest

========================================
BuildTest
========================================
Executing task: BuildTest
Analyzing build script...
Processing build script...
Installing tools...
Compiling build script...
**Script compile time: 2584ms**
**MSBuild.ExtensionPack_version=4.0.15.cake execution time: 3534ms**
Deleting file D:/CacheTest/Basic_Recipe/MSBuild.ExtensionPack_version=4.0.15.cake
Build
Finished executing task: BuildTest

========================================
CleanProject
========================================
Executing task: CleanProject
Finished executing task: CleanProject

----------------------------------------
Teardown
----------------------------------------
Executing custom teardown action...
Starting Teardown...
Cleaning Project Complete
Deleting nupkg files...
Finished running tasks.

Task                          Duration
--------------------------------------------------
Setup                         00:00:00.0126425
Hello                         00:00:00.0121108
VersionTest                   00:00:04.5714435
BuildTest                     00:00:04.4771106
CleanProject                  00:00:00.0038407
Teardown                      00:00:00.0527423
--------------------------------------------------
Total:                        00:00:09.1298904
**build.cake execution time: 12544ms**

(Last Command - 16 seconds, 506 milliseconds)

Preparing to run build script...
Running Cake bootstrap...
Running build script...
Analyzing build script...
Processing build script...
Installing addins...
build.cake.dll
Running cached build script...

----------------------------------------
Setup
----------------------------------------
Executing custom setup action...
  ____         _           _____             _
 / ___|  __ _ | | __  ___ |_   _|  ___  ___ | |_
| |     / _` || |/ / / _ \  | |   / _ \/ __|| __|
| |___ | (_| ||   < |  __/  | |  |  __/\__ \| |_
 \____| \__,_||_|\_\ \___|  |_|   \___||___/ \__|


Starting Setup...
Cleaning Project

========================================
Hello
========================================
Executing task: Hello
Hello
Finished executing task: Hello

========================================
VersionTest
========================================
Executing task: VersionTest
Analyzing build script...
Processing build script...
Installing tools...
GitVersion.CommandLine_version=5.0.1.cake.dll
**Running cached build script...**
**GitVersion.CommandLine_version=5.0.1.cake execution time: 9ms**
Deleting file D:/CacheTest/Basic_Recipe/GitVersion.CommandLine_version=5.0.1.cake
Version
Finished executing task: VersionTest

========================================
BuildTest
========================================
Executing task: BuildTest
Analyzing build script...
Processing build script...
Installing tools...
MSBuild.ExtensionPack_version=4.0.15.cake.dll
**Running cached build script...**
**MSBuild.ExtensionPack_version=4.0.15.cake execution time: 8ms**
Deleting file D:/CacheTest/Basic_Recipe/MSBuild.ExtensionPack_version=4.0.15.cake
Build
Finished executing task: BuildTest

========================================
CleanProject
========================================
Executing task: CleanProject
Finished executing task: CleanProject

----------------------------------------
Teardown
----------------------------------------
Executing custom teardown action...
Starting Teardown...
Cleaning Project Complete
Deleting nupkg files...
Finished running tasks.

Task                          Duration
--------------------------------------------------
Setup                         00:00:00.0154086
Hello                         00:00:00.0106217
VersionTest                   00:00:00.9394647
BuildTest                     00:00:00.9310428
CleanProject                  00:00:00.0040268
Teardown                      00:00:00.0535480
--------------------------------------------------
Total:                        00:00:01.9541126
**build.cake execution time: 2020ms**

(Last Command - 5 seconds, 895 milliseconds)

File: src/Cake.Core/Constants.cs [hunk 1]
```diff
@@ -19,11 +19,17 @@ public static class Settings
             public const string ShowProcessCommandLine = "Settings_ShowProcessCommandLine";
         }
 
+        public static class Cache
```
Comment (mholo65) [inline src/Cake.Core/Constants.cs]: As Cache is a part of Cake, not Cake.Core, these constants should be moved there.

File: src/Cake.Core/Extensions/CakeConfigurationExtensions.cs [hunk 2]
```diff
@@ -43,5 +43,23 @@ public static DirectoryPath GetModulePath(this ICakeConfiguration configuration,
...
         }
+
+        /// <summary>
+        /// Gets the script cache directory path.
+        /// </summary>
+        /// <param name="configuration">The Cake configuration.</param>
+        /// <param name="defaultRoot">The default root path.</param>
+        /// <param name="environment">The environment.</param>
+        /// <returns>The script cache directory path.</returns>
+        public static DirectoryPath GetScriptCachePath(this ICakeConfiguration configuration, DirectoryPath defaultRoot, ICakeEnvironment environment)
```
Comment (mholo65) [inline src/Cake.Core/Extensions/CakeConfigurationExtensions.cs]: Same as with constants. Move this extension to Cake.

File: src/Cake.Core/Utilities/DisposableStopwatch.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+using System.Diagnostics;
+
+using Cake.Core.Diagnostics;
+
+namespace Cake.Core.Utilities
+{
+    /// <summary>
+    /// Disposable Stopwatch used for timing how long a section of code takes and writing the result to the log.
+    /// </summary>
+    public class DisposableStopwatch : IDisposable
```
Comment (mholo65) [inline src/Cake.Core/Utilities/DisposableStopwatch.cs]: Don't think we need a stopwatch at this point. Please open separate issue for reporting compilation time.

File: src/Cake.Core/Utilities/FastHash.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Cake.Core.Utilities
+{
+    /// <summary>
+    /// Optimized hash generator. Using SHA1CryptoServiceProvider since it is FIPS compliant.
+    /// </summary>
+    public static class FastHash
```
Comment (mholo65) [inline src/Cake.Core/Utilities/FastHash.cs]: Hashing is only needed for Caching at this point, move to Cake.

File: src/Cake/Arguments/ArgumentParser.cs [hunk 5]
```diff
@@ -187,6 +187,11 @@ private bool ParseOption(string name, string value, CakeOptions options)
                 options.Exclusive = ParseBooleanValue(value);
             }
 
+            if (name.Equals("recompile", StringComparison.OrdinalIgnoreCase))
```
Comment (mholo65) [inline src/Cake/Arguments/ArgumentParser.cs]: Suggestion for better name? `invalidate-cache` or similar. Something that refers to cache would be good.

File: src/Cake/Scripting/Roslyn/RoslynScriptEngine.cs [hunk 6]
```diff
@@ -11,19 +12,21 @@ namespace Cake.Scripting.Roslyn
     internal sealed class RoslynScriptEngine : IScriptEngine
     {
         private readonly CakeOptions _options;
+        private readonly ICakeConfiguration _configuration;
         private readonly IAssemblyLoader _loader;
         private readonly ICakeLog _log;
 
-        public RoslynScriptEngine(CakeOptions options, IAssemblyLoader loader, ICakeLog log)
+        public RoslynScriptEngine(CakeOptions options, ICakeConfiguration configuration, IAssemblyLoader loader, ICakeLog log)
```
Comment (mholo65) [inline src/Cake/Scripting/Roslyn/RoslynScriptEngine.cs]: @patriksvensson if I remember correctly, you where removing places where we pass `IConfiguration` around in your CLI rewrite PR?

Comment (tstewart65) [reply src/Cake/Scripting/Roslyn/RoslynScriptEngine.cs]: Yes this part is not needed when using the CLI review RFC0001.  I have also made the changes for the cache using the RFC0001 pull request.

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs [hunk 7]
```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
...
-                        break;
-                    case DiagnosticSeverity.Error:
-                        _log.Error(diagnostic.ToString());
-                        errors.Add(diagnostic);
-                        break;
-                    default:
-                        break;
+                    _log.Verbose(cacheDLLFileName);
+                    scriptHash = FastHash.GenerateHash(Encoding.UTF8.GetBytes(string.Concat(script.Lines)));
+                    var cachedHash = IO.File.Exists(hashFile.FullPath) ? IO.File.ReadAllText(hashFile.FullPath) : string.Empty;
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Don't use `System.IO` use the `IFileSystem`

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs [hunk 8]
```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
...
+                    {
+                        var errorMessages = string.Join(Environment.NewLine, errors.Select(x => x.ToString()));
+                        var message = string.Format(CultureInfo.InvariantCulture, "Error(s) occurred when compiling build script:{0}{1}", Environment.NewLine, errorMessages);
+                        throw new CakeException(message);
+                    }
+                }
+                if (_scriptCacheEnabled)
+                {
+                    // Verify cache directory exists
+                    if (!IO.Directory.Exists(_scriptCachePath.FullPath))
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Don't use `System.IO` use the `IFileSystem`

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs [hunk 9]
```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
...
+                    {
+                        IO.Directory.CreateDirectory(_scriptCachePath.FullPath);
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Don't use `System.IO` use the `IFileSystem`

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs [hunk 10]
```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
...
-            if (errors.Any())
-            {
-                var errorMessages = string.Join(Environment.NewLine, errors.Select(x => x.ToString()));
-                var message = string.Format(CultureInfo.InvariantCulture, "Error(s) occurred when compiling build script:{0}{1}", Environment.NewLine, errorMessages);
-                throw new CakeException(message);
-            }
-
+        private void RunScriptAssembly(string assemblyPath)
+        {
+            var assembly = Assembly.LoadFile(assemblyPath);
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Use `IAssemblyLoader` here

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs [hunk 11]
```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
...
+            var type = assembly.GetType("Submission#0");
+            var factoryMethod = type.GetMethod("<Factory>", new[] { typeof(object[]) });
             using (new ScriptAssemblyResolver(_log))
             {
-                roslynScript.RunAsync(_host).Wait();
+                try
+                {
+                    var task = (System.Threading.Tasks.Task<object>)factoryMethod.Invoke(null, new object[] { new object[] { _host, null } });
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: No need for `System.Threading.Tasks`. Include the namespace instead.

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs [hunk 12]
```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
...
+                    }
+                    if (string.IsNullOrEmpty(scriptHash))
+                    {
+                        scriptHash = FastHash.GenerateHash(Encoding.UTF8.GetBytes(string.Concat(script.Lines)));
+                    }
+                    var emitResult = compilation.Emit(cachedAssembly.FullPath);
+
+                    if (emitResult.Success)
+                    {
+                        IO.File.WriteAllText(hashFile.FullPath, scriptHash);
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Don't use `System.IO` use the `IFileSystem`

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs [hunk 13]
```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
...
+                    }
+                }
+                if (_scriptCacheEnabled)
+                {
+                    // Verify cache directory exists
+                    if (!IO.Directory.Exists(_scriptCachePath.FullPath))
+                    {
+                        IO.Directory.CreateDirectory(_scriptCachePath.FullPath);
+                    }
+                    if (string.IsNullOrEmpty(scriptHash))
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Use `IsNullOrWhitespace`

**Review by patriksvensson — DISMISSED** [review]

File: src/Cake.Core/Constants.cs (hunk 1 above)
Comment (tstewart65) [reply src/Cake.Core/Constants.cs]: Moved to Cake/Constants.cs

File: src/Cake.Core/Extensions/CakeConfigurationExtensions.cs (hunk 2 above)
Comment (tstewart65) [reply src/Cake.Core/Extensions/CakeConfigurationExtensions.cs]: Moved to Cake/Extensions/CakeConfiguration.cs

File: src/Cake.Core/Utilities/DisposableStopwatch.cs (hunk 3 above)
Comment (tstewart65) [reply src/Cake.Core/Utilities/DisposableStopwatch.cs]: Removed

File: src/Cake.Core/Utilities/FastHash.cs (hunk 4 above)
Comment (tstewart65) [reply src/Cake.Core/Utilities/FastHash.cs]: Moved to Cake/Utilities/FastHash.cs

File: src/Cake/Arguments/ArgumentParser.cs (hunk 5 above)
Comment (tstewart65) [reply src/Cake/Arguments/ArgumentParser.cs]: Change "recompile" to "regeneratecache"

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs (hunk 7 above)
Comment (tstewart65) [reply src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Fixed

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs (hunk 8 above)
Comment (tstewart65) [reply src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Fixed

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs (hunk 9 above)
Comment (tstewart65) [reply src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Fixed

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs (hunk 10 above)
Comment (tstewart65) [reply src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Fixed

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs (hunk 11 above)
Comment (tstewart65) [reply src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Fixed

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs (hunk 12 above)
Comment (tstewart65) [reply src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Fixed

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs (hunk 13 above)
Comment (tstewart65) [reply src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Fixed

**Comment by pascalberger** [comment]
I updated the code to the changes made in #3381 and rebased on latest `develop` branch

File: src/Cake/Commands/DefaultCommandSettings.cs [hunk 14]
```diff
@@ -59,5 +59,9 @@ public sealed class DefaultCommandSettings : CommandSettings
         [CommandOption("--info")]
         [Description("Displays additional information about Cake.")]
         public bool ShowInfo { get; set; }
+
+        [CommandOption("--regeneratecache")]
```
Comment (bjorkstromm) [inline src/Cake/Commands/DefaultCommandSettings.cs:63]: maybe `--invalidate-cache`?

File: src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs [hunk 15]
```diff
@@ -48,6 +60,11 @@ public sealed class RoslynScriptSession : IScriptSession
             ReferencePaths = new HashSet<FilePath>(PathComparer.Default);
             References = new HashSet<Assembly>();
             Namespaces = new HashSet<string>(StringComparer.Ordinal);
+
+            var cacheEnabled = configuration.GetValue(Constants.Cache.Enabled) ?? "false";
```
Comment (bjorkstromm) [inline src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs:64]: `bool.FalseString`

File: src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs [hunk 16]
```diff
@@ -48,6 +60,11 @@ public sealed class RoslynScriptSession : IScriptSession
...
+            _scriptCacheEnabled = cacheEnabled.Equals("true", StringComparison.OrdinalIgnoreCase);
```
Comment (bjorkstromm) [inline src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs:65]: `bool.TrueString`

File: src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs [hunk 17]
```diff
@@ -48,6 +60,11 @@ public sealed class RoslynScriptSession : IScriptSession
...
+            _regenerateCache = host.Context.Arguments.HasArgument("regeneratecache");
```
Comment (bjorkstromm) [inline src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs:66]: `invalidate-cache`

Comment (bjorkstromm) [reply src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs:66]: Would also be good to avoid "magic strings". Do we have a `Constants.cs` where these should reside?

File: src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs [hunk 18]
```diff
@@ -82,6 +99,30 @@ public void ImportNamespace(string @namespace)
...
+            var hashFile = _scriptCachePath.CombineWithFilePath(cacheHashFileName);
+            string scriptHash = default;
+            if (_scriptCacheEnabled && _fileSystem.Exist(cachedAssembly) && !_regenerateCache)
+            {
+                _log.Verbose($"Cache enabled: Checking cache build script ({cacheDLLFileName})");
+                scriptHash = FastHash.GenerateHash(Encoding.UTF8.GetBytes(string.Concat(script.Lines)));
+                var cachedHash = _fileSystem.Exist(hashFile)
+                               ? _fileSystem.GetFile(hashFile).ReadLines(Encoding.UTF8).FirstOrDefault()
+                               : string.Empty;
+                if (scriptHash.Equals(cachedHash, StringComparison.InvariantCultureIgnoreCase))
```
Comment (bjorkstromm) [inline src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs:115]: `OrdinalIgnoreCase` or just`Ordinal`?

File: src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs [hunk 19]
```diff
@@ -159,7 +200,49 @@ public void Execute(Script script)
...
+
+        private void RunScriptAssembly(string assemblyPath)
+        {
+            var assembly = _loader.Load(assemblyPath, false);
+            var type = assembly.GetType("Submission#0");
+            var factoryMethod = type.GetMethod("<Factory>", new[] { typeof(object[]) });
+            try
+            {
+                var task = (Task<object>)factoryMethod.Invoke(null, new object[] { new object[] { _host, null } });
+                task.Wait();
```
Comment (bjorkstromm) [inline src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs:240]: If we'd use `GetAwaiter().GetResult()` instead of `Wait()`, we could probably skip the `try / catch` which rethrows inner exception.

File: src/Cake/Utilities/FastHash.cs [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    internal static class FastHash
+    {
+        /// <summary>
+        /// Generates a hash of the passed byte arrays.
+        /// </summary>
+        /// <param name="input">The binary data to hash.</param>
+        /// <returns>The hash value.</returns>
+        public static string GenerateHash(byte[] input)
+        {
+            using (var sha1 = new SHA1CryptoServiceProvider())
```
Comment (bjorkstromm) [inline src/Cake/Utilities/FastHash.cs:24]: Should we use `SHA256` or `SHA512`?

File: src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs (hunk 18 above)
Comment (devlead) [reply src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs:115]: FastHash.GenerateHash will always have the same casing and characters, so Ordinal should work.

File: src/Cake/Utilities/FastHash.cs (hunk 20 above)
Comment (devlead) [reply src/Cake/Utilities/FastHash.cs]: Yes, `sha1` will cause compliance issues,  also providers are now obsolete so the base Create method must be used.

I've pushed a commit that uses `SHA512` and new more performant `Convert.ToHexString` on .NET 5 / 6.

File: src/Cake/Commands/DefaultCommandSettings.cs (hunk 14 above)
Comment (devlead) [reply src/Cake/Commands/DefaultCommandSettings.cs:63]: @bjorkstromm maybe `--force-cache-recompile` and `"Forces script to recompiled and cached (if caching is enabled)."`

?

File: src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs (hunk 19 above)
Comment (devlead) [reply src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs]: Adjusted.

**Review by nils-a — APPROVED** [review]
LGTM

File: src/Cake.Core/Constants.cs [hunk 21]
```diff
@@ -27,5 +27,12 @@ public static class Paths
             public const string Addins = "Paths_Addins";
             public const string Modules = "Paths_Modules";
         }
+
+        public static class Cache
+        {
+            public const string Enabled = "Cache_Enabled";
+            public const string Path = "Cache_Path";
```
Comment (nils-a) [inline src/Cake.Core/Constants.cs:34]: This could, however, also be `Paths_Cache` - to have the paths in one place...

**Comment by nils-a** [comment]
I created https://github.com/cake-build/resources/issues/104 to keep track of the need to update the `cake.config` in https://github.com/cake-build/resources

File: src/Cake.Core/Constants.cs (hunk 21 above)
Comment (devlead) [reply src/Cake.Core/Constants.cs]: @nils-a adjusted to `Settings.EnableScriptCache` and `Paths.Cache` so. command-line becomes  i.e

`--settings_enablescriptcache=true --paths_cache="C:/temp/cache"`

**Review by bjorkstromm — COMMENTED** [review]
I'd like to see all changes made in `Cake.Core` moved out to `Cake`. Not a hill I'm prepared to die on, but I'd argue it makes more sense there as script cache, and compilation is something that happens in `Cake`, and not in `Cake.Core`.

**Comment by nils-a** [comment]
@bjorkstromm are you proposing we move the Configuration parts from `Cake.Core` into `Cake`? IMHO this would better be an individual PR.

**Comment by bjorkstromm** [comment]
> @bjorkstromm are you proposing we move the Configuration parts from `Cake.Core` into `Cake`? IMHO this would better be an individual PR.

No, I'm only proposing the additions in this PR (one extension method, and two constants) would be better suited living in Cake than Cake.Core.

**Comment by devlead** [comment]
> > @bjorkstromm are you proposing we move the Configuration parts from `Cake.Core` into `Cake`? IMHO this would better be an individual PR.
> 
> No, I'm only proposing the additions in this PR (one extension method, and two constants) would be better suited living in Cake than Cake.Core.

Yip makes sense to me. Agree 👍

**Review by bjorkstromm — APPROVED** [review]
LGTM!

**Review by nils-a — APPROVED** [review]

**Comment by devlead** [comment]
@tstewart65  your changes have been merged, thanks for your contribution 👍

**Comment by GeertvanHorrik** [comment]
First of all, amazing work and I am really happy to see this being built into Cake.

We use the same scripts for many of our repositories, but would like to cache this in a generic way on our build agents. For example, we have 150 repositories that would be using this feature and it would be hard to explicitly set a different path for each of them in cake.config.

I checked the source, but could not find an easy answer to whether this will cache in the repository itself or in the temp path. If it's in the temp path, how will it differentiate between different working directories, etc?

Basically what we are looking for is how we can cache this:

1. Unique cache per repository (so we need some sort of key, but the paths will be different per build)
2. Cache inside %temp%, not inside the repository since that will be cleared on each build

If this all works as I hope, this will shave up 4 minutes of each build since we have 6 stages and each of them need to (re)compile the scripts right now which takes up about 45 seconds (large scripts).

**Comment by devlead** [comment]
@GeertvanHorrik by default it will end up in a cache folder within the default [tools folder](https://cakebuild.net/docs/running-builds/configuration/default-configuration-values#tools-path).

But cache location fully configurable
* Config
  * Paths.Cache = /path/to/cache
* Arg
  * --paths_cache=/path/to/cache
* Env
  * CAKE_PATHS_CACHE=/path/to/cache

**Comment by GeertvanHorrik** [comment]
Thanks for the fast reply. We even templated our cake.config. Would it be possible to use variables or how does it determine the final subfolder inside the cache.

For example (and I am happy to contribute), something like this:

- Cache path is set to %temp%\cake-scripts-cache\
- Subdirectory is a hash of all the script files to be compiled, so ends up in

%temp%\cake-scripts-cache\[some-unique-hash]

I see 2 things that would be needed here:

1. a lock file, so if 2 builds are running at the same time, cake knows not to use the directory (yet). If the lock file is older than 5 minutes, assume it's broken and restart
2. a hash calculation (fastest possible, but it should be fairly fast since I don't expect more than 50 files to be used for a script)

**Comment by GeertvanHorrik** [comment]
Thanks. I think that I need 1 more sublevel for the cache (otherwise it will keep being invalidated), but maybe I should try the feature first and only if it turns out to be an issue, force the cache directory per repository.

Looking forward to try this one out, I assume it will be part of Cake 2.2? Any ETA or pre-release available?

**Comment by devlead** [comment]
Pre-releases are always available via our public Azure Artifacts feed
* https://dev.azure.com/cake-build/Cake/_packaging?_a=package&feed=cake&package=Cake.Tool&protocolType=NuGet

2.2 is preliminarily planned to be released this or next week.

As you're using `cake.config`, would something like the below config, enabling env var expansion solve your concerns?

```ini
[Paths]
Cache=%TEMP%/tools
```

**Comment by GeertvanHorrik** [comment]
I would need 1 more level in caching, e.g.:

[Paths]
Cache=%TEMP%/cake-build/%myrepo%

I am probably overthinking and should just take the cake.config out of the repository template.

**Comment by devlead** [comment]
@GeertvanHorrik That actually already works, I'll pr an integration test to demonstrate and fix a typo in this one.

**Comment by GeertvanHorrik** [comment]
That is perfect, then I can't wait to try the new version and can keep cake.config in my repository synchronizer template.
